daemon/logger/awslogs/cloudwatchlogs.go

```diff
@@ -140,19 +143,54 @@ func New(info logger.Info) (logger.Logger, error) {
 	if err != nil {
 		return nil, err
 	}
+
 	containerStream := &logStream{
 		logStreamName:    logStreamName,
 		logGroupName:     logGroupName,
 		logCreateGroup:   logCreateGroup,
+		logNonBlocking:   logNonBlocking,
 		multilinePattern: multilinePattern,
 		client:           client,
 		messages:         make(chan *logger.Message, 4096),
 	}
-	err = containerStream.create()
-	if err != nil {
-		return nil, err
+
+	creationDone := make(chan bool)
+	if logNonBlocking {
+		go func() {
+			backoff := 1
+			maxBackoff := 32
+			for {
+				// If logger is closed we are done
+				containerStream.lock.RLock()
+				if containerStream.closed {
+					containerStream.lock.RUnlock()
+					break
+				}
+				containerStream.lock.RUnlock()
+				err := containerStream.create()
+				if err == nil {
+					break
+				}
+
+				time.Sleep(time.Duration(backoff) * time.Second)
```

Comment:
This is a good question.

The classes of errors I can think of all seem to have the ability to heal over time. The classes I can think of are:

* Network availability 
* Service errors 
* Service Limits 
* Credentials/policy issue

Do you see something outside of these classes or a way that one of these won't be able to possibly heal?

Reply:
IMO, Credentials and policy issues dont heal over time. They need manual intervention to fix the issue. Also service errors is a broad category. There's a possibility that there are some types of service errors that dont auto-heal with time.

Reply:
Credential and policy issues can be healed out-of-band by changes in the IAM policy.  They won't automatically heal, but can be fixed while the container is running.

By "service errors", @IRCody is talking about the kinds of errors that would be fixed by AWS, also out-of-band with respect to the container.  Network availability and service errors are the general issue that we're trying to address with this code.

Reply:
Sounds good.

Reply:
Agree, all things that can be fixed outside of the scope of the container (e.g. when error logs start showing up).
On a separate note it might be nice (in a another exercise) to track metrics for errors here.